### ISSUE TITLE
404 error on "<yarp>/systemweb-adapters/authenticate?original-url=%2FTicket" when trying to go to Core page

### ISSUE THREAD

**Issue by JS-Nati** [dotnet/systemweb-adapters#320]
**_Is there a better or more appropriate place for me to ask about this kind of assistance?_**

TLDR explanation: 
We're getting an error, and I can't seem to see why or what configuration I'm missing, between documentation and samples...
Before we started implementing the auth/sessoin sharing we were able to run both sites together, you just had to log into them each independently. We are trying to implement session/auth sharing to fix the UX issues with having to log in over and over again as they transition between old and new pages. Now I can log into Framework, but when I try to go to a Core page it gives 404.

Full explanation:
I'm working on an update to .NET 6 of an old legacy ASP.NET Framework project. Previously we'd updated it to run in Framework 4.7.2, so that's not an issue at all... We jumped on board VERY early with the SystemWebAdapters, and ended up missing out on the auth/session sharing setup. We had the yarp working fine, but the logins between old/new were really annoying in our testing, so we know it'll be an issue for our users. Well I'm to a point where I'm trying to use the auth/session, and I'm at a very frustrating error.

After finally getting to the point where I'm using the Ver. 1.1.0 of the DLLs (we were using the oldest available nuget package) I've worked my way through to the point where I CAN log into the Framework site, and the YARP is working for the Framework pages, but when I try to open a page in core it's giving me a 404 error (According to Fiddler the missing path is: /systemweb-adapters/authenticate?original-url=%2FTicket).

**Code:**
For Core I've left a lot of commented code in place because we had already implemented Identity and Session in Core at this point and I'm not 100% sure it all needs to be commented for this to work.
Global.Asax.cs where I'm setting up the adapters (We do have custom code for Session_Start and other things also, but don't know if that is relevant):
```c#
protected void Application_Start() {
	FilterConfig.RegisterGlobalFilters( GlobalFilters.Filters );
	AreaRegistration.RegisterAllAreas();
	RouteTable.Routes.MapRoute(<customRoute>);
	RouteConfig.RegisterRoutes( RouteTable.Routes );
	using( var context = new UsersContext() ) {
		context.Database.Initialize( true );
	}
	BootstrapBundleConfig.RegisterBundles( System.Web.Optimization.BundleTable.Bundles );
	ValueProviderFactories.Factories.Remove( ValueProviderFactories.Factories.OfType<System.Web.Mvc.JsonValueProviderFactory>().FirstOrDefault() );
	ValueProviderFactories.Factories.Add( new OurCode.App_Start.MyJsonValueProviderFactory() );
	AntiForgeryConfig.SuppressIdentityHeuristicChecks = true;

	SystemWebAdapterConfiguration.AddSystemWebAdapters( this )
		.AddProxySupport( options => options.UseForwardedHeaders = true )
		.AddJsonSessionSerializer( options => {
			options.RegisterKey<SiteData>( Globals.SessionKey.Site );
			options.RegisterKey<CurrentUser>( Globals.SessionKey.User );
		} )
		.AddRemoteAppServer( options => options.ApiKey = ConfigurationManager.AppSettings["RemoteAppApiKey"] )
		.AddAuthenticationServer()
		.AddSessionServer();
}

``` 

And here's my Program.cs:

```c#
var builder = WebApplication.CreateBuilder(args);


//We implemented Core Identity here, but I commented it out to try and use the SystemWebAdapters Auth sharing

#region Config settings
//We implement our custom config options/objects here
#endregion

#region System Web Adapters for reverse proxy
builder.Services.AddSystemWebAdapters()
	//Setup Session sharing
	.AddJsonSessionSerializer( options => {
		// Serialization/deserialization requires each session key to be registered to a type
		options.RegisterKey<SiteData>( Globals.SessionKey.Site );
		options.RegisterKey<UserData>( Globals.SessionKey.User );
	} )
	.AddRemoteAppClient( options => {
		// Provide the URL for the remote app that has enabled session querying
		options.RemoteAppUrl = new( conf["ReverseProxy:Clusters:fallbackCluster:Destinations:fallbackApp:Address"] );

		options.ApiKey = conf["RemoteAppApiKey"];
	} )
	.AddAuthenticationClient( true )
	.AddSessionClient();
builder.Services.AddReverseProxy().LoadFromConfig( conf.GetSection( "ReverseProxy" ) );

#endregion

// Add services to the container.
builder.Services.AddControllersWithViews();
//builder.Services.AddRazorPages();

//builder.Services.ConfigureApplicationCookie( options =>
//{
//	options.LoginPath = $"/coreaccount/login";
//	options.LogoutPath = $"/coreaccount/logout";
//} );

//Enable Core session (Used before trying to use SystemWebAdapters
//builder.Services.AddDistributedMemoryCache();
//builder.Services.AddSession( options => {
//	options.IdleTimeout = TimeSpan.FromMinutes( 10 );
//    options.Cookie.HttpOnly = true;
//	options.Cookie.IsEssential = true;
//} );


var app = builder.Build();

// Configure the HTTP request pipeline.
if (!app.Environment.IsDevelopment())
{
    app.UseExceptionHandler("/Home/Error");
    // The default HSTS value is 30 days. You may want to change this for production scenarios, see https://aka.ms/aspnetcore-hsts.
    app.UseHsts();
}

app.UseHttpsRedirection();
app.UseStaticFiles();

app.UseRouting();
app.UseAuthentication();
app.UseAuthorization();
//app.UseSession(); //put back to use core session
app.UseSystemWebAdapters();
//Enable and use both routing by convention and attribute routing
app.MapControllerRoute(
	name: "default",
	pattern: "{controller=Home}/{action=Index}/{id?}" ).RequireSystemWebAdapterSession();

//app.MapRazorPages();
app.MapReverseProxy();
//Used to insert our middleware to make sure session variables were populated before trying to use SystemWebAdapters session sharing
//app.ApplyOurMiddleware(); // Implement custom middleware

app.Run();
```

**Comment by twsouthwick** [comment]
@JS-Nati Hmm, I'll see what I can do to help you here, but a full repro if possible would be really helpful.

A couple of thoughts:

- The 404 you are seeing to `/systemweb-adapters/authenticate?original-url=%2FTicket` is the authentication attempting to contact the framework server to get the ClaimsPrincipal. What kind of authentication are you using? We have samples in the repo for examples on how to set up for various common auth scenarios
- Is the end user error a 404? I'd expect it to be a 500 as the 404 would be expected given the URL for the auth endpoint, but would translate into a 500 on the core app

**Comment by JS-Nati** [comment]
I'll do my best to address your 2 points:

First: I do not know a lot about the how of it, as at least 3 previous devs built the Framework code before I came into the picture, but I do know that we're overriding with our own CustomPrincipal object that inherits ICustomPrincipal and we handle login/verification with custom code also, though it seems to adhere to the basic process of Framework Identity, as it upgraded relatively easily to Core Identity, with only a few new columns, and it uses standard decorators for controller/action auth. On the core side we're doing very little custom, other than pointing to our tables, to get Identity working. From what I can tell there's no customization of "ClaimsPrincipal" on the framework side (I can't find any references in the code to "claimsprincipal" or "claims").

As far as a duplication, I'm not sure I can provide that as I am not 100% how to duplicate without just putting our code out, which isn't something I can do. I do understand that duplication would be useful, I just don't currently have a way to strip things down and still know that it'll duplicate our side without compromising our code. I do know that, though we have some custom action filters we ARE using standard authorize decorators and code mechanisms on the framework side. We haven't overrided Auth, just login.

Is there a simple way/place for me to put a breakpoint and look for objects that can help to understand if our custom code is missing something the adapters are expecting?

Second point: The error definitely STARTED as a backend 404 on the authenticate yarp route that ended in a browser display of 502 server error, after a bunch of failed 404 yarp authenticate calls. The current configuration, what I shared, is actually showing 404 at the browser, and I've been fiddling with things so much that I'm not 100% sure exactly what I changed to get it from the 502 to what's currently happening, but it's no longer failing on the authenticate repeatedly, it's a single call. I'm guessing it's now a 404 because it's not actually authorising and so the Core call comes up as 404, the same as it would if core identity authorize failed. 

I've downloaded the code, zip file, from the git repo and between documentation and the samples configured our code to what I shared. It seems to me that whatever conditions are being used to decide on making the adapters' route, or possibly something odd about how we're handling routing for our Framework MVC, the adapters route isn't able to be generated, or something along those lines. That's all assuming I don't have anything egregiously wrong/bad about how I've implemented the doc/samples in our code.

Here's the fiddler info for the core page and corresponding the yarp call, sans app key and truncated .ASPXAUTH=<AuthData> I've also attached a full fiddler session from debug start to Ticket yarp failure: (lots of adapter authenticate 404s)
[26_Full_AppId removed.txt](https://github.com/dotnet/systemweb-adapters/files/11191776/26_Full_AppId.removed.txt)


```
GET https://localhost:7138/Ticket HTTP/1.1
Host: localhost:7138
Connection: keep-alive
sec-ch-ua: "Google Chrome";v="111", "Not(A:Brand";v="8", "Chromium";v="111"
sec-ch-ua-mobile: ?0
sec-ch-ua-platform: "Windows"
Upgrade-Insecure-Requests: 1
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/111.0.0.0 Safari/537.36
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7
Sec-Fetch-Site: same-origin
Sec-Fetch-Mode: navigate
Sec-Fetch-User: ?1
Sec-Fetch-Dest: document
Referer: https://localhost:7138/AdminAccount/GetSiteList
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.9
Cookie: ASP.NET_SessionId=1fjpqh3ejhltfqg4c5rvxk0b; .ASPXAUTH=<AuthData>; __RequestVerificationToken=49kvMA74_7J4-k0frC5nY7QFl0OocP_f1FBzSAjBATl-yQyOoW64bTpykzy71ueKRFdydcdCQmvRwi5Q_97y349K9ZnrTfHpAE2yEeHFXTU1


HTTP/1.1 404 Not Found
Content-Length: 0
Date: Fri, 07 Apr 2023 16:56:29 GMT
Server: Kestrel



------------------------------------------------------------------

GET http://localhost:52524/systemweb-adapters/authenticate?original-url=%2FTicket HTTP/1.1
Host: localhost:52524
x-forwarded-host: localhost:7138
x-forwarded-proto: https
x-migration-authenticate: true

Cookie: ASP.NET_SessionId=1fjpqh3ejhltfqg4c5rvxk0b; .ASPXAUTH=<AuthData>; __RequestVerificationToken=49kvMA74_7J4-k0frC5nY7QFl0OocP_f1FBzSAjBATl-yQyOoW64bTpykzy71ueKRFdydcdCQmvRwi5Q_97y349K9ZnrTfHpAE2yEeHFXTU1
X-SystemWebAdapter-RemoteAppAuthentication-Key: <GUID>
traceparent: 00-146937f902a647fea5b538f2f92c95f5-949f440197dea6a9-00


HTTP/1.1 404 Not Found
Cache-Control: private
Content-Type: text/html; charset=utf-8
Server: Microsoft-IIS/10.0
X-AspNet-Version: 4.0.30319
X-SourceFiles: =?UTF-8?B?RTpcc291cmNlXHJlcG9zXEJpdGJ1Y2tldFxEZXZDb3JlXE9ydGhydXNXZWJcT3J0aHJ1c1dlYlxPcnRocnVzXHN5c3RlbXdlYi1hZGFwdGVyc1xhdXRoZW50aWNhdGU=?=
X-Powered-By: ASP.NET
Date: Fri, 07 Apr 2023 16:56:30 GMT
Content-Length: 2914

<!DOCTYPE html>
<html>
    <head>
        <title>The resource cannot be found.</title>
        <meta name="viewport" content="width=device-width" />
        <style>
         body {font-family:"Verdana";font-weight:normal;font-size: .7em;color:black;} 
         p {font-family:"Verdana";font-weight:normal;color:black;margin-top: -5px}
         b {font-family:"Verdana";font-weight:bold;color:black;margin-top: -5px}
         H1 { font-family:"Verdana";font-weight:normal;font-size:18pt;color:red }
         H2 { font-family:"Verdana";font-weight:normal;font-size:14pt;color:maroon }
         pre {font-family:"Consolas","Lucida Console",Monospace;font-size:11pt;margin:0;padding:0.5em;line-height:14pt}
         .marker {font-weight: bold; color: black;text-decoration: none;}
         .version {color: gray;}
         .error {margin-bottom: 10px;}
         .expandable { text-decoration:underline; font-weight:bold; color:navy; cursor:pointer; }
         @media screen and (max-width: 639px) {
          pre { width: 440px; overflow: auto; white-space: pre-wrap; word-wrap: break-word; }
         }
         @media screen and (max-width: 479px) {
          pre { width: 280px; }
         }
        </style>
    </head>

    <body bgcolor="white">

            <span><H1>Server Error in '/' Application.<hr width=100% size=1 color=silver></H1>

            <h2> <i>The resource cannot be found.</i> </h2></span>

            <font face="Arial, Helvetica, Geneva, SunSans-Regular, sans-serif ">

            <b> Description: </b>HTTP 404. The resource you are looking for (or one of its dependencies) could have been removed, had its name changed, or is temporarily unavailable. &nbsp;Please review the following URL and make sure that it is spelled correctly.
            <br><br>

            <b> Requested URL: </b>/systemweb-adapters/authenticate<br><br>

            <hr width=100% size=1 color=silver>

            <b>Version Information:</b>&nbsp;Microsoft .NET Framework Version:4.0.30319; ASP.NET Version:4.8.4494.0

            </font>

    </body>
</html>
<!-- 
[HttpException]: The controller for path &#39;/systemweb-adapters/authenticate&#39; was not found or does not implement IController.
   at System.Web.Mvc.DefaultControllerFactory.GetControllerInstance(RequestContext requestContext, Type controllerType)
   at System.Web.Mvc.MvcHandler.ProcessRequestInit(HttpContextBase httpContext, IController& controller, IControllerFactory& factory)
   at System.Web.Mvc.MvcHandler.BeginProcessRequest(HttpContextBase httpContext, AsyncCallback callback, Object state)
   at System.Web.HttpApplication.CallHandlerExecutionStep.System.Web.HttpApplication.IExecutionStep.Execute()
   at System.Web.HttpApplication.ExecuteStepImpl(IExecutionStep step)
   at System.Web.HttpApplication.ExecuteStep(IExecutionStep step, Boolean& completedSynchronously)
-->

------------------------------------------------------------------


```

**Comment by JS-Nati** [comment]
OK, So I was able to track down what was happening that caused the systemweb-adapters/authenticate and session routes not to load, causing a 404 on the core side.

Because of the age and/or complexity of our Framework setup the Module section for the adapters wasn't added to the web.config.
I added the below section to system.webServer and those routes are loaded now:
```c#
<modules runAllManagedModulesForAllRequests="true">
	<remove name="SystemWebAdapterModule"/>
	<add name="SystemWebAdapterModule"
		type="Microsoft.AspNetCore.SystemWebAdapters.SystemWebAdapterModule, Microsoft.AspNetCore.SystemWebAdapters.FrameworkServices"
		preCondition="managedHandler"/>
</modules>
``` 

This fixed the issue of the Framework side 404 error and loaded the modules for Authorization and Session. Now I'm having issues with our CustomPrincipal and authentication. We are using a "GenericIdentity" in our CustomPrincipal and so the only thing that **was** coming across to Core was a ClaimsPrincipal with a username and nothing else.

I have modified our old code to add role claims, verified in debug, but I'm still getting 403-Forbidden when I try to hit Core routes. I thought that role based authorization would just need the username and role claims. 

On the Core side of our code we're not using any custom claims/principal objects, we've only modified the code to point to our app DB, tables, and columns (so that old and new can use the same table(s)). Other than that we were using standard Identity for .NET 6. I've commented this config out above, but it's standard code with DI objects to point the Identity User object to some of our old tables/columns for the login/data.

```c#
//start: Core Identity setup
//var connectionString = builder.Configuration.GetConnectionString("OrthruswebCoreContextConnection") ?? throw new InvalidOperationException("Connection string 'OrthruswebCoreContextConnection' not found.");

//builder.Services.AddDbContext<OrthruswebCoreContext>(options =>
//    options.UseSqlServer(connectionString));

//builder.Services.AddDefaultIdentity<OrthruswebCoreUser>(options => options.SignIn.RequireConfirmedAccount = true)
//    .AddRoles<OrthruswebCoreRole>()
//    .AddEntityFrameworkStores<OrthruswebCoreContext>()
//    .AddUserManager<OrthruswebCoreUserManager>()
//    .AddDefaultUI()
//    .AddDefaultTokenProviders();
//end: Core Identity setup
``` 

Before I go digging into SystemWebAdapters code and settings from the Git repo again to try and find what I'm not seeing/understanding, do you have any ideas as to why I would be having issues with the Identity sent from SystemWebAdapters that has a list of role claims and a username for [Authorize] controllers?

I'm sure it's something stupid that I'm just overlooking.

**Comment by JS-Nati** [comment]
Just realized I didn't tag @twsouthwick on my replies, above.

Please let me know what you think. I'm going to get together some samples of our customPrincipal object, which is being cast to ClaimsPrincipal, and is causing our current authorization issue.

Here's code for CustomPrincipal that is being cast to ClaimsPrincipal to send back to Core. (ICustomPrincipal is just our custom data that inherits from IPrincipal)
```c#
public class CustomPrincipal : ICustomPrincipal
{
    public IIdentity Identity { get; private set; }
    public bool IsInRole(string role) 
    {
        return this.Roles.Contains(role, StringComparer.OrdinalIgnoreCase);
    }

    public bool HasRoleIn(params string[] roles)
    {
        foreach(var v in roles)
        {
            if ( IsInRole( v ) ) return true;
        }
        return false;
    }

    public CustomPrincipal(string email)
    {
        this.Identity = new GenericIdentity(email);
    }

    <We have a bunch of custom data for our users here as simple properties>
}
``` 

I've modified this object in a couple ways trying to get the object on the other side to allow access to the Core pages, but Identity.IsAuthorized=false every time. I adjusted the Roles to add the claims, but most notably I tried modifying the constructor setting the identity to be:
```c#
this.Identity = new GenericIdentity(email,"Remote");
``` 
I did this because I'm trying to figure out how to get the Core side Identity.IsAuthenticated to be true. Also I figured that role based authorization would require the roles claims.

**Comment by JS-Nati** [comment]
I'm going to close this question, as I've already handled the original issue.

Below is review of the solution from above:

I was able to track down what was happening that caused the systemweb-adapters/authenticate and session routes not to load, causing a 404 on the core side.

Because of the age and/or complexity of our Framework setup the Module section for the adapters wasn't added to the web.config.
I added the below section to system.webServer and those routes are loaded now:
```c#
<modules runAllManagedModulesForAllRequests="true">
	<remove name="SystemWebAdapterModule"/>
	<add name="SystemWebAdapterModule"
		type="Microsoft.AspNetCore.SystemWebAdapters.SystemWebAdapterModule, Microsoft.AspNetCore.SystemWebAdapters.FrameworkServices"
		preCondition="managedHandler"/>
</modules>
``` 
This fixed the issue of the Framework side 404 error and loaded the modules for Authorization and Session.
